File: backend/app/schemas/billing.py
"""Billing & subscription Pydantic schemas."""

from datetime import datetime

from pydantic import BaseModel


# ── Plan Info ──


class PlanFeatures(BaseModel):
    price: int | None  # CNY fen, None = custom
    posts: int  # -1 = unlimited
    platforms: int
    lines: int
    competitors: int
    publish_trial_days: int | None = None


class PlanInfo(BaseModel):
    name: str
    label: str
    features: PlanFeatures
    is_current: bool = False


# ── Subscription ──


class SubscriptionResponse(BaseModel):
    id: int
    plan: str
    status: str
    trial_ends_at: datetime | None
    current_period_start: datetime
    current_period_end: datetime
    monthly_post_limit: int
    platform_limit: int
    content_line_limit: int
    competitor_account_limit: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SubscribeRequest(BaseModel):
    plan: str  # free/basic/pro/enterprise


class UpgradeRequest(BaseModel):
    new_plan: str


# ── Usage ──


class UsageSummary(BaseModel):
    posts_generated: int
    posts_published: int
    post_limit: int  # -1 = unlimited
    platforms_used: list[str]
    platform_limit: int
    content_lines_used: list[str]
    content_line_limit: int
    period_start: datetime
    period_end: datetime
    usage_percent: float  # 0-100 based on posts


# ── Overage ──


class OverageLineItem(BaseModel):
    resource: str  # posts / platforms / lines
    used: int
    limit: int
    overage: int
    unit_price_cents: int
    total_cents: int


class OverageBill(BaseModel):
    items: list[OverageLineItem]
    total_cents: int
    currency: str = "CNY"


# ── Payment ──


class PaymentIntent(BaseModel):
    payment_id: str
    amount_cents: int
    currency: str = "CNY"
    method: str
    payment_url: str  # WeChat Pay / Alipay redirect
    expires_at: datetime


class PaymentResponse(BaseModel):
    id: int
    amount_cents: int
    currency: str
    method: str
    status: str
    description: str | None
    paid_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class PayRequest(BaseModel):
    amount_cents: int
    method: str = "wechat_pay"
    description: str | None = None


# ── Usage Limit Check Result ──


class UsageLimitResult(BaseModel):
    allowed: bool
    reason: str | None = None
    current: int
    limit: int  # -1 = unlimited

File: frontend/src/services/billing.ts
import { api } from './api'

// ── Types ──

export interface PlanFeatures {
  price: number | null
  posts: number
  platforms: number
  lines: number
  competitors: number
  publish_trial_days: number | null
}

export interface PlanInfo {
  name: string
  label: string
  features: PlanFeatures
  is_current: boolean
}

export interface SubscriptionResponse {
  id: number
  plan: string
  status: string
  trial_ends_at: string | null
  current_period_start: string
  current_period_end: string
  monthly_post_limit: number
  platform_limit: number
  content_line_limit: number
  competitor_account_limit: number
  created_at: string
  updated_at: string
}

export interface UsageSummary {
  posts_generated: number
  posts_published: number
  post_limit: number
  platforms_used: string[]
  platform_limit: number
  content_lines_used: string[]
  content_line_limit: number
  period_start: string
  period_end: string
  usage_percent: number
}

export interface OverageLineItem {
  resource: string
  used: number
  limit: number
  overage: number
  unit_price_cents: number
  total_cents: number
}

export interface OverageBill {
  items: OverageLineItem[]
  total_cents: number
  currency: string
}

export interface PaymentIntent {
  payment_id: string
  amount_cents: number
  currency: string
  method: string
  payment_url: string
  expires_at: string
}

export interface PaymentRecord {
  id: number
  amount_cents: number
  currency: string
  method: string
  status: string
  description: string | null
  paid_at: string | null
  created_at: string
}

// ── API Functions ──

export async function getSubscription() {
  const { data } = await api.get<SubscriptionResponse>('/v1/billing/subscription')
  return data
}

export async function subscribe(plan: string) {
  const { data } = await api.post<SubscriptionResponse>('/v1/billing/subscribe', { plan })
  return data
}

export async function upgradePlan(newPlan: string) {
  const { data } = await api.post<SubscriptionResponse>('/v1/billing/upgrade', {
    new_plan: newPlan,
  })
  return data
}

export async function cancelSubscription() {
  const { data } = await api.post<SubscriptionResponse>('/v1/billing/cancel')
  return data
}

export async function getUsageSummary() {
  const { data } = await api.get<UsageSummary>('/v1/billing/usage')
  return data
}

export async function getOverage() {
  const { data } = await api.get<OverageBill>('/v1/billing/overage')
  return data
}

export async function getPlans() {
  const { data } = await api.get<PlanInfo[]>('/v1/billing/plans')
  return data
}

export async function createPayment(amountCents: number, method = 'wechat_pay', description?: string) {
  const { data } = await api.post<PaymentIntent>('/v1/billing/pay', {
    amount_cents: amountCents,
    method,
    description,
  })
  return data
}

export async function getPaymentHistory() {
  const { data } = await api.get<PaymentRecord[]>('/v1/billing/payments')
  return data
}
